src/app/replayer/sql.ml

```diff
@@ -0,0 +1,202 @@
+(* sql.ml -- (Postgresql) SQL queries for replayer *)
+
+open Core_kernel
+
+(* build query to find all blocks back to genesis block, starting with the block containing the
+   specified state hash; for each such block, find ids of all (user or internal) commands in that block
+*)
+
+let find_command_ids_query s =
+  sprintf
+    {|
+      WITH RECURSIVE chain AS (
+
+        SELECT id,parent_id FROM blocks b WHERE b.state_hash = ?
+
+        UNION ALL
+
+        SELECT b.id,b.parent_id FROM blocks b
+
+        INNER JOIN chain
+
+        ON b.id = chain.parent_id
+      )
+
+      SELECT DISTINCT %s_command_id FROM chain c
+
+      INNER JOIN
+
+      blocks_%s_commands bc
+
+      ON bc.block_id = c.id
+
+     |}
+    s s
+
+module User_command_ids = struct
+  let query =
+    Caqti_request.collect Caqti_type.string Caqti_type.int
+      (find_command_ids_query "user")
+
+  let run (module Conn : Caqti_async.CONNECTION) state_hash =
+    Conn.collect_list query state_hash
+end
+
+module User_command = struct
+  type t =
+    { type_: string
+    ; fee_payer_id: int
+    ; source_id: int
+    ; receiver_id: int
+    ; fee: int64
+    ; fee_token: int64
+    ; token: int64
+    ; amount: int64 option
+    ; nonce: int64
+    ; global_slot: int64
+    ; sequence_no: int }
+
+  let typ =
+    (* chunk into groups so we can use tuple combinators *)
+    let encode t =
+      Ok
+        ( (t.type_, t.fee_payer_id, t.source_id, t.receiver_id)
+        , (t.fee, t.fee_token, t.token, t.amount)
+        , (t.nonce, t.global_slot, t.sequence_no) )
+    in
+    let decode
+        ( (type_, fee_payer_id, source_id, receiver_id)
+        , (fee, fee_token, token, amount)
+        , (nonce, global_slot, sequence_no) ) =
+      Ok
+        { type_
+        ; fee_payer_id
+        ; source_id
+        ; receiver_id
+        ; fee
+        ; fee_token
+        ; token
+        ; amount
+        ; nonce
+        ; global_slot
+        ; sequence_no }
+    in
+    let rep =
+      Caqti_type.(
+        tup3 (tup4 string int int int)
+          (tup4 int64 int64 int64 (option int64))
+          (tup3 int64 int64 int))
+    in
+    Caqti_type.custom ~encode ~decode rep
+
+  (* do we need to consider fee_payer/receiver_account creation fee paid? *)
```

Comment:
We shouldn't need to: replaying the transaction should be deterministically the same, and that's where we infer this extra information from initially.